### ISSUE TITLE
Live binding not ready? Solved with nested imports?

### ISSUE THREAD

**Issue by trusktr** [webpack/webpack#2735]
For example, I have two modules, A and B that are a circular dependency:

```js
// B.js
import A, {defineClassA} from `./A`

if (!A) defineClassA() // this fires, and triggers the error, see A.js.

export default
class B extends A { /* ... */ }
```

```js
// A.js
import someFunction from './other-module'
import B from './B'

let A = null
export default A

if (!A) defineClassA() // this doesn't fire, the error is triggered by B.js first.

export
function defineClassA() {
  someFunction() // Error happens here, triggered from B.js

  A = class {
    method() {
      console.log('B constructor:', B)
    }
  }
}
```

The error is

```
Uncaught TypeError: Cannot read property 'default' of undefined
```

where the compiled code looks like

```js
(0, _someFunction2.default)();
```

There seems to be a module load-order issue at play. Any recommendations on how to handle this? I could just add the `import someFunction` statement into B.js, which solves the problem, but seems awkward since I'm not using `someFunction` inside of B.js.

This also leads me to think that nested imports would force a solution to the problem here. See [reify](https://github.com/benjamn/reify) and [The case for nested `import` declarations](https://github.com/benjamn/reify/blob/master/WHY_NEST_IMPORTS.md). cc @benjamn

**Comment by trusktr** [comment]
Update, importing `someFunction` (note, I'm using generic names, my case has other names, but is very similar) does not solve the problem, which leads me to think that there might be a live-binding issue. For example, in B.js `someFunction` can be observed to exist by logging it, yet the same error still happens in A.js:

```js
// B.js
import someFunction from './other-module'
import A, {defineClassA} from `./A`

console.log(someFunction) // it exists, so the live binding in A should also exist then?

if (!A) defineClassA() // this fires, and triggers the error, so the "live binding" in A is still undefined?

export default
class B extends A { /* ... */ }
```

**Comment by trusktr** [comment]
The reason I am having this problem is because I was trying to solve the circular dependency. Class A needs to be defined in order to define class B, but the B module is evaluated first (I'm not exactly sure what defines the load order), so I thought I could solve the problem by wrapping the `class A` definition in a function and then executing that in B.

Is there another way?

**Comment by trusktr** [comment]
Oops, forgot to import B in A.js showing the circular dependency, I've updated the example.

**Comment by trusktr** [comment]
I've found a very hacky workaround by passing the module dependencies to the `defineClassA` function:

```js
// B.js
import someFunction from './other-module'
import A, {defineClassA} from `./A`

if (!A) defineClassA(someFunction) // <-- how ugly!

export default
class B extends A { /* ... */ }
```

```js
// A.js
import someFunction from './other-module'
import B from './B'

let A = null
export default A

if (!A) defineClassA(someFunction)

export
function defineClassA(someFunction) {
  someFunction() // Error happens here, triggered from B.js

  A = class {
    method() {
      console.log('B constructor:', B)
    }
  }
}
```

Hideous, but it works. Any other ways to do it?

**Comment by trusktr** [comment]
Apparently the live binding idea (to solve the need for class A to be defined for class B in this circular dependency) didn't work, as after the `defineClassA` call, `A` is still undefined:

```js
// B.js
import someFunction from './other-module'
import A, {defineClassA} from `./A`

if (!A) defineClassA(someFunction) // <-- this should work, since it modifies the live binding from A.js

console.log(A) // undefined! But shouldn't be!

export default
class B extends A { /* ... */ }
```

and I get the error

```
Uncaught TypeError: Super expression must either be null or a function, not undefined        inherits.js:21
```

which is the error I was trying to solve all along with the original code where `class A` where the original code was

```js
// B.js
import A from `./A`

export default
class B extends A { /* ... */ }
```

```js
// A.js
import someFunction from './other-module'
import B from './B'

someFunction()

export default
class A {
  method() {
    console.log('B constructor:', B)
  }
}
```

It's obvious to see in this last example that A will be undefined in B.js because only one module can be executed at once.

**Comment by trusktr** [comment]
TLDR: In theory, I think live bindings allow a solution to this circular dependency problem, as it would be impossible if the variables were limited to the module scopes. But, in this actual case with Webpack, it seems like I may have found a corner case where "live bindings" aren't working as I thought?

**Comment by trusktr** [comment]
@benjamn When testing in Meteor, the console.log statement is also undefined, and I get the same error (worded differently):

```
Uncaught TypeError: Super expression must either be null or a function, not undefined
```

Maybe I'm missing something about how live bindings work? I thought if I modified the binding as I've done with `defineClassA` that it would then be available by the time I define class B.

**Comment by sokra** [comment]
I think changing `export default A` to `export { A as default }` will solve your issue.

Note that `export default A` does NOT bind A as default export. `export default` takes an expression and not an reference. The expression is only evaluated once. While `export { A as default }` binds A as default export.

Take a look at the code babel generates for the different types:

## [export default](https://babeljs.io/repl/#?evaluate=true&lineWrap=false&presets=es2015%2Creact%2Cstage-2&experimental=false&loose=true&spec=false&code=let%20A%20%3D%201%3B%0A%0Aexport%20default%20A%0A%0AA%20%3D%202%3B%0A)

``` js
let A = 1;

export default A

A = 2;
```

``` js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var A = 1;

exports.default = A;


A = 2;
```

## [export { A as default }](https://babeljs.io/repl/#?evaluate=true&lineWrap=false&presets=es2015%2Creact%2Cstage-2&experimental=false&loose=true&spec=false&code=let%20A%20%3D%201%3B%0A%0Aexport%20%7B%20A%20as%20default%20%7D%0A%0AA%20%3D%202%3B%0A)

``` js
let A = 1;

export { A as default }

A = 2;
```

``` js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var A = 1;

exports.default = A;


exports.default = A = 2;`
``

**Comment by trusktr** [comment]
There doesn't seem to be live binding in either of those two example modules, as when a third module imports A modifies the value, the new value won't propagate into these two exporting modules.

On top of that, in my case, the value is `undefined` although I'm exporting a class. For example,

```js
export default class Foo {}
```

compared to

```js
class Foo {}
export {Foo as default}
```

In either case, I would not expect an `undefined` value when importing from either module. I have a dependency loop, maybe that's causing the issue somehow.

**Comment by trusktr** [comment]
Hmmm, after investigating, it does seem like a dependency loop causing the problem. Trying to see if I can eliminate it...

**Comment by trusktr** [comment]
Interesting to note that Meteor seems to have a similar problem with my code base when I run it there. (@benjamn).

**Comment by trusktr** [comment]
Possibly related reproduction: run the script from https://github.com/webpack/webpack/issues/2740#issuecomment-230879978, then in `webpack-issue-2740/app/src/app/index.js` toggle the comments on lines 8 through 11, from

```js
// import Scene from 'infamous/motor/Scene'
// console.log('Scene?', Scene)
import Node from 'infamous/motor/Node'
console.log('Node:', Node)
```

to

```js
import Scene from 'infamous/motor/Scene'
console.log('Scene?', Scene)
// import Node from 'infamous/motor/Node'
// console.log('Node:', Node)
```

This will cause the module evaluation order to be different, in which case the `Node` variable is `undefined`.

**Comment by trusktr** [comment]
FWIW, the problem I had _was_ due to circular dependencies. [This](https://esdiscuss.org/topic/how-to-solve-this-basic-es6-module-circular-dependency-problem) ES Discuss thread details how to solve circular dependency requirements using "init functions" that are hoisted and can be executed by modules other than the init-function-defining module before that module is evaluated.

**Comment by trusktr** [comment]
I also feel that the difference between `export default A` and `export {A as default}` is going to cause headaches for many years.
